File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Array/Recursions/ArrayRecursion.py
# Recursion in Array - Recursive search in inventory (MotorPH)


def recursive_search(inventory, brand, index=0):
    """Search inventory array by brand using recursion."""
    if index >= len(inventory):
        return None
    if inventory[index]["brand"] == brand:
        return inventory[index]
    return recursive_search(inventory, brand, index + 1)


def display(inventory):
    """Display inventory list."""
    if not inventory:
        print("  (empty)")
        return
    for item in inventory:
        print(f"  - {item['brand']} {item['model']}: {item['quantity']} units")


def main():
    # Inventory array
    inventory = [
        {"brand": "Honda", "model": "Click 125", "quantity": 10},
        {"brand": "Yamaha", "model": "NMAX 155", "quantity": 5},
        {"brand": "Suzuki", "model": "Burgman Street", "quantity": 7},
    ]

    print("Inventory:")
    display(inventory)

    # Recursive search by brand
    print("\nRecursive search for 'Yamaha':")
    result = recursive_search(inventory, "Yamaha")
    print(f"  {result if result else 'Not found'}")

    print("\nRecursive search for 'Honda':")
    result = recursive_search(inventory, "Honda")
    print(f"  {result if result else 'Not found'}")

    print("\nRecursive search for 'Kawasaki':")
    result = recursive_search(inventory, "Kawasaki")
    print(f"  {result if result else 'Not found'}")


if __name__ == "__main__":
    main()

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Array/Recursions/ArrayRecursion.java
package Array.Recursions;

import java.util.ArrayList;
import java.util.List;

/**
 * Recursion in Array - Recursive search in inventory (MotorPH)
 */
public class ArrayRecursion {

    record InventoryItem(String brand, String model, int quantity) {}

    /** Search inventory array by brand using recursion. */
    static InventoryItem recursiveSearch(List<InventoryItem> inventory, String brand, int index) {
        if (index >= inventory.size()) {
            return null;
        }
        if (inventory.get(index).brand().equals(brand)) {
            return inventory.get(index);
        }
        return recursiveSearch(inventory, brand, index + 1);
    }

    /** Display inventory list. */
    static void display(List<InventoryItem> inventory) {
        if (inventory.isEmpty()) {
            System.out.println("  (empty)");
            return;
        }
        for (InventoryItem item : inventory) {
            System.out.println("  - " + item.brand() + " " + item.model() + ": " + item.quantity() + " units");
        }
    }

    public static void main(String[] args) {
        // Inventory array (ArrayList)
        List<InventoryItem> inventory = new ArrayList<>();
        inventory.add(new InventoryItem("Honda", "Click 125", 10));
        inventory.add(new InventoryItem("Yamaha", "NMAX 155", 5));
        inventory.add(new InventoryItem("Suzuki", "Burgman Street", 7));

        System.out.println("Inventory:");
        display(inventory);

        // Recursive search by brand
        System.out.println("\nRecursive search for 'Yamaha':");
        InventoryItem result = recursiveSearch(inventory, "Yamaha", 0);
        System.out.println("  " + (result != null ? result : "Not found"));

        System.out.println("\nRecursive search for 'Honda':");
        result = recursiveSearch(inventory, "Honda", 0);
        System.out.println("  " + (result != null ? result : "Not found"));

        System.out.println("\nRecursive search for 'Kawasaki':");
        result = recursiveSearch(inventory, "Kawasaki", 0);
        System.out.println("  " + (result != null ? result : "Not found"));
    }
}
